Introduce distributed checkpoint with ShardedTensor.

kumpera opened the pull request:
Co-authored-by: Wen Zhang <zhangwen@fb.com>
Co-authored-by: Yifu Wang <yifu@fb.com>



cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu

kumpera committed:
Introduce distributed checkpoint with ShardedTensor.

Co-authored-by: Wen Zhang <zhangwen@fb.com>
Co-authored-by: Yifu Wang <yifu@fb.com>

zzzwen commented:
Thanks for the PR. qq, what is the naming conversation for prototype feature in test folder?  I notice that it is "_checkpoint" in the src but "checkpoint" in the test.

Since we have been through the review cycle for this prototype feature already, shall we stamp and merge this

kumpera commented:
> Thanks for the PR. qq, what is the naming conversation for prototype feature in test folder? I notice that it is "_checkpoint" in the src but "checkpoint" in the test.
> 
> Since we have been through the review cycle for this prototype feature already, shall we stamp and merge this

Tests are not part of the API so it's ok to land them `checkpoint` and I don't think there will be any confusion on how it's supposed to work.

While we need this stamped, it can't be merged before our tests are green.

pritamdamania87 commented on torch/distributed/_checkpoint/state_dict_loader.py:
```diff
@@ -0,0 +1,234 @@
+        state_dict: A state_dict to verify if it's loadable.
+        metadata: Checkpoint metadata to verify against.
+
+    Returns:
+        None if no issue was found or a List[str] of issues.
```
We should probably just throw an appropriate RuntimeError instead of returning a string of errors like this.

pritamdamania87 commented on test/distributed/checkpoint/test_file_system_checkpoint.py:
```diff
@@ -0,0 +1,490 @@
+# Owner(s): ["oncall: distributed"]
```
Awesome tests!

pritamdamania87 commented:
Awesome work @kumpera!

kumpera commented on torch/distributed/_checkpoint/metadata.py:
```diff
@@ -0,0 +1,92 @@
+@dataclass
+class ExtendedTensorMetadata:
+    # Details of Tensor/ShardedTensor (dtype, shape, sharding config etc.)
+    # TODO: It might not make sense to force Tensor's metadata into
+    # ShardedTensorMetadata, let's create a metadata class for regular tensor
+    tensor_metadata: ShardedTensorMetadata
```
Good point. It actually simplifies the code as the current metadata we need to store for Tensor is significantly smaller.

kumpera commented on torch/distributed/_checkpoint/metadata.py:
```diff
@@ -0,0 +1,92 @@
+        self.state_dict_metadata = pickle.loads(state)
+
+
+@dataclass
+class BytesWriteRequest:
+    bytes: io.BytesIO
+    storage_key: str
+
+
+@dataclass
+class BytesReadRequest:
+    bytes: io.BytesIO
+    storage_key: str
+
+
+@dataclass
+class TensorWriteRequest:
+    tensor: torch.Tensor
+    storage_key: str
+
+
+@dataclass
+class TensorReadRequest:
```
Yes, there was a long discussion around non-Tensor and ByteIO seems to be the most reasonable choice that still enables a working default implementation.
We need to properly document that decision and follow up with validation by using an alternative serialization format.

kumpera commented on torch/distributed/_checkpoint/resharding.py:
```diff
@@ -0,0 +1,197 @@
+            assert shard_md_from_storage is not None
+            # this is a naive quadratic algo that can later be optimized
+            #   by sorting metadata and the shards md
+            # FIXME how do we handle offset > 0?
```
Because there's no place to pass it down.
The idea was that this would be used by the storage layer when bundling multiple tensors in a single file and this would be the offset into that file.

I removed this as I don't believe it's the right design.
This information can easily be kept by the storage layer itself.

kumpera commented on torch/distributed/_checkpoint/state_dict_loader.py:
```diff
@@ -0,0 +1,234 @@
+    """
+    Use the loaded metadata and the current state dict to map the saved tensors to current tensors
+
+    NOTE:
+    During the save,
```
Yes, but there's no trace of what that would be. Just removing it

kumpera commented on torch/distributed/_checkpoint/state_dict_loader.py:
```diff
@@ -0,0 +1,234 @@
+    tensor_read_requests = []
+    bytes_read_requests = []
+    for fqn, obj in state_dict.items():
+        if isinstance(obj, torch.Tensor):
```
Won't this work just fine with either as they have shape and dtype set?

kumpera commented on torch/distributed/_checkpoint/state_dict_loader.py:
```diff
@@ -0,0 +1,234 @@
+    for fqn, obj in state_dict.items():
+        if isinstance(obj, torch.Tensor):
+            tensor = obj.detach()
+            storage_size = tensor.nelement() * tensor.element_size()
```
Good point.

kumpera commented on torch/distributed/_checkpoint/state_dict_loader.py:
```diff
@@ -0,0 +1,234 @@
+    storage_reader: StorageReader,
+) -> None:
+    """
+    This public function defines the default behavior to load a state_dict
```
Updated, we should probably write more general notes into a module doc page like this one: https://pytorch.org/docs/stable/_sources/distributed.rst.txt

kumpera replied:
https://github.com/pytorch/pytorch/issues/76323

kumpera commented on torch/distributed/_checkpoint/state_dict_saver.py:
```diff
@@ -0,0 +1,189 @@
+    storage_writer: StorageWriter,
+) -> None:
+    """
+    This public function defined the default behavior to save a state_dict
```
I think there's too much to be documented on a single method and it's better to have package level documentation for this feature. Tracking this here: https://github.com/pytorch/pytorch/issues/76323

kumpera commented on torch/distributed/_checkpoint/state_dict_saver.py:
```diff
@@ -0,0 +1,189 @@
+    """
+    This public function defined the default behavior to save a state_dict
+    Notes
+    1. This is a WIP, the state_dict save with different versions of the code might not be compatible.
```
good point.

kumpera commented on torch/distributed/_checkpoint/state_dict_saver.py:
```diff
@@ -0,0 +1,189 @@
+    This public function defined the default behavior to save a state_dict
+    Notes
+    1. This is a WIP, the state_dict save with different versions of the code might not be compatible.
+    2. The caller needs to ensure the correctness of the state_dict
```
AFAICT, right now this is only about calling `_register_state_dict_hook`.
I changed the wording to reflect that  which is actionable

zzzwen replied:
maybe we should change the comment, my point was that the code does not guarantee whether the state_dict it receive is a correct/complete representation of the nn.module

kumpera commented on torch/distributed/_checkpoint/storage_reader.py:
```diff
@@ -0,0 +1,96 @@
+            # narrow it along all dimemsions, and copy_ it to the
+            # target tensor, which will be the same size.
+            for dim, (start, length) in enumerate(zip(req.offsets, req.lengths)):
+                view_to_copy = torch.narrow(view_to_copy, dim, start, length)
```
This issue is not as serious as it looks since we only keep one tensor in memory at a time.

We decided in favor of this tradeoff as it's enables us to support arbitrary tensors while the alternative would require to effectively reimplement a large part of torch.load / torch.save.

Tracking it here: https://github.com/pytorch/pytorch/issues/76331

zzzwen reviewed (commented):
Looks like we want to make the first commit more mature. so, channel some of the requirements I get from the chat with @simpkins .

Another important one is the storage read/write should not have knowledge on communication and the torch.distributed. I think we can address it in the next PR

kumpera commented on torch/distributed/_checkpoint/state_dict_saver.py:
```diff
@@ -0,0 +1,189 @@
+    Uses the state_dict to build three things.
+
+    metadata: Metadata
+        The metatdata discribing the tensor / sharded tensor.
```
I reviewed the whole doc string for this function

kumpera commented on torch/distributed/_checkpoint/state_dict_saver.py:
```diff
@@ -0,0 +1,189 @@
+        The metatdata discribing the tensor / sharded tensor.
+        And it is storage meta data. See "../metadata.py" for detail
+
+    size_for_storage_keys: Dict[str, int]
+        Key is the storage key name, value is its size
+        It can used to pre allocate the storage for parallel and non sequential writes.
```
The Storage Write has a method `prepare_storage` that uses this information to possibly prime plan / reserve storage space.

This dict features rank specific requests, so it's not easy to derive it from metadata, but trivial from the tensor write requests.

@zzzwen could you better explain the value of `prepare_storage`?

zzzwen replied:
size_for_storage_keys might be an over design; the size can be figured out with metadata manually.
However, IMO, the prepare_storage can be essential, most storage need initialization one way or another, and the initialization is likely required to be performed outside of init()

pritamdamania87 replied:
In that case, we can just have `prepare_storage` take the metadata as a parameter instead?

kumpera replied:
I think we should just remove prepare_storage as everything it does can be done in write_tensors at no loss of performance.

kumpera commented on torch/distributed/_checkpoint/state_dict_saver.py:
```diff
@@ -0,0 +1,189 @@
+    bytes_write_requests: List[BytesWriteRequest]
+        List of byte write requests that should br perfromed by the writer.
+
+    Subclasses can optionally overwrite the implementation here,
+    if the default does not meet its requirement.
```
We removed this extension point until we figure out how we want it done.

kumpera commented on torch/distributed/_checkpoint/state_dict_saver.py:
```diff
@@ -0,0 +1,189 @@
+    if the default does not meet its requirement.
+    Args:
+        state_dict: The state_dict to operate on
+        always_add_tensors: Include non-sharded tensors even if rank != 0
```
It's for testing. This allows to write tests that don't require to be on rank0.

pritamdamania87 replied:
I don't think we should add flags like this in the code just for testing purposes. We can test this accordingly in unit tests by doing things like `if self.rank == 0: validate_something()`. I feel the latter is more robust since otherwise your unit tests are testing a codepath that is never actually triggered in practice.

kumpera commented on torch/distributed/_checkpoint/state_dict_saver.py:
```diff
@@ -0,0 +1,189 @@
+                tensor_write_requests.append(
+                    TensorWriteRequest(
+                        tensor=obj.detach(),
+                        storage_key=fqn,
```
Yes, but `always_add_tensor=True` is currently only used for testing and without `dist.init`.

kumpera commented on torch/distributed/_checkpoint/state_dict_saver.py:
```diff
@@ -0,0 +1,189 @@
+                        storage_key=fqn,
+                    )
+                )
+                metadata.state_dict_metadata[fqn] = _compute_tensor_md(fqn, obj)
```
I updated the docstring to mention that this argument must not be used with load/save.

kumpera commented on torch/distributed/_checkpoint/state_dict_saver.py:
```diff
@@ -0,0 +1,189 @@
+            )
+            bytes_write_requests.append(bwr)
+
+    storage_keys: Dict[str, int] = {
+        req.storage_key: req.tensor.nelement() * req.tensor.element_size()
+        for req in tensor_write_requests
+    }
```
While this is true, the StorageWriter contract says that we pass this `dict` to `prepare_storage` before issuing the any writes, so we'd need to change it.

wanchaol commented on test/distributed/checkpoint/test_file_system_checkpoint.py:
```diff
@@ -0,0 +1,496 @@
+        model_to_load = MyShardedModel3(dst_spec)
+        model_to_load._register_state_dict_hook(state_dict_hook)
+        state_dict_to_load_to = model_to_load.state_dict()
+
```
don't we need a barrier here? what if some rank (i.e. rank 0) finishs saving the state_dict and some not, then rank 0 starts loading the state_dict but rank 1 still not yet finishs saving the local shards to the directory? This might still work with no-resharding, but with resharding, I think it might cause some problem?

kumpera replied:
I'm not sure what the barrier would be coordinating here.

Load-time resharding is a pure-local operation where each rank independently plans how to load from the checkpoint.

kumpera replied:
Filed a follow up issue on ensuring load/save has reasonable SPMD coordination behavior

kumpera replied:
https://github.com/pytorch/pytorch/issues/76395

wanchaol commented on torch/distributed/_checkpoint/state_dict_loader.py:
```diff
@@ -0,0 +1,250 @@
+                    f"Invalid checkpoint metadata for {fqn}, " +
+                    "expected TensorStorageMetadata but found {type(md)}"
+                )
+        elif isinstance(obj, ShardedTensor):
```
same here, let's do `isinstance` check for ShardedTensor first then torch.Tensor second to avoid issues when torch.Tensor subclass PR landed

wanchaol commented on torch/distributed/_checkpoint/state_dict_saver.py:
```diff
@@ -0,0 +1,150 @@
+                    )
+                )
+                metadata.state_dict_metadata[fqn] = _compute_tensor_md(fqn, obj)
+        elif isinstance(obj, ShardedTensor):
```
same here for `isinstance` check

wanchaol reviewed (commented):
Great work! Some suggestions and some questions inlined.

kumpera commented on torch/distributed/_checkpoint/storage_reader.py:
```diff
@@ -0,0 +1,98 @@
+    """
+
+    @abc.abstractmethod
+    def read_bytes(self, requests: List[BytesReadRequest]) -> Future[None]:
```
I think unless we know what that something clearly is (regardless of the type) we should not change the signature.

Since those values would be implementation dependent on the subclass of StorageReader, users could just `cast(Future[Something], reader.read_bytes(...))`.

zzzwen replied:
IIRC,  Future[Any]  does not pass the linter

kumpera commented on torch/distributed/_checkpoint/state_dict_loader.py:
```diff
@@ -0,0 +1,250 @@
+    state_dict: Dict[str, Any],
+    storage_reader: StorageReader,
+    dont_read_tensors: bool = False,
+    read_but_not_copy: bool = False,
```
Those two are dead code I introduced as part of my debug-by-PR despair. I'm removing them as soon as I figure out the crash.

kumpera commented on torch/distributed/_checkpoint/resharding.py:
```diff
@@ -0,0 +1,196 @@
+            ):
+                # Note that we do NOT want to make any tensor copy.
+                # all operation must be view only
+                target_tensor = torch.narrow(
```
What we do here is create a view to where in the tensor we want data loaded to. The tensor from `narrow` shares storage with the original tensor.

The actual copy happens in FileSystemReader::read_tensors. See do `req.tensor.copy_(view_to_copy)` where `req.tensor` is the previously narrow'd tensor and `view_to_copy` has the data.

kumpera commented on torch/distributed/_checkpoint/metadata.py:
```diff
@@ -0,0 +1,92 @@
+# everything is fluid and can be changed
+#
+# 1. We will need a metadata class for regular tensor, no point forcing it
+# into a ShardedTensorMetadata.
```
Done it. There's `ShardedTensorStorageMetadata` and `TensorStorageMetadata`

kumpera commented on torch/distributed/_checkpoint/resharding.py:
```diff
@@ -0,0 +1,197 @@
+        # For most cases, the read is a recovery from a failure to the same sharding
+        # and does not need any resharding, write each shard as is is the most effective
+        suffix = "_".join([str(i) for i in shard.metadata.shard_offsets])
+        storage_key = f"{storage_key_prefix}_{suffix}"
```
Changed the separator to`$` and introduced `resharding.STORAGE_KEY_SEPARATOR` constant to track that.
I refactored this into `_get_shard_storage_key` to avoid duplication.

kumpera commented on torch/distributed/_checkpoint/storage_writer.py:
```diff
@@ -0,0 +1,107 @@
+
+
+class FileSystemWriter(StorageWriter):
+    def __init__(self, path: str) -> None:
```
We should resolve this in a follow up PR

kumpera commented:
[linux-xenial-py3.7-gcc5.4 / test (backwards_compat, 1, 1, linux.2xlarge)](https://github.com/pytorch/pytorch/runs/6182626428?check_suite_focus=true#logs) failure is unrelated.

kumpera commented:
@wanchaol  @pritamdamania87 I believe I addressed all review issues, broken tests.
There's a lot of work ahead beyond this prototype, I filed half a dozen issues on it under `sharded_tensor`

simpkins commented on torch/distributed/_shard/checkpoint/filesystem.py:
```diff
@@ -0,0 +1,127 @@
+
+    def write_bytes(self, requests: List[BytesWriteRequest]) -> Future[None]:
+        for req in requests:
+            with open(os.path.join(self.path, req.storage_key), "wb") as storage:
```
Especially for the case of handling case-insensitive file systems, maybe it would be better to not use the key at all in the path name.  It seems like it would perhaps be better to just pick arbitrary path names, and have a mapping of storage key to path in the metadata.

This would also make it easier to only save 1 copy of the data if `save_state_dict()` is called with multiple tensors that all point to the same storage.

zzzwen replied:
Great point, the handle/file name hashing is essential for production system. I
IUC, we are all talking about doing the name hash inside the StorageWriter/Reader but not above it, right? 
And I assume the comment here is more about the advanced/production case and not about this basic OSS file system.

zzzwen replied:
> On your last point, about having multiple tensors pointing to the same storage, how common is to run into models where multiple tensors cover overlapping parts of their underlying storage?

For checkpointing case, I tend to consider it a bug in state_dict. Shall we create a utility function that checks overlapping memory, in state_dict_saver/state_dict_loader, and prevent it from reaching the storagerwriter/reader

simpkins replied:
See T87045442 for one example where we have models with shared storage in the state dict.

I also think we should build a clean API here that makes no assumptions about how it is being used.  I think we should provide a generic key-value store, that doesn't care what the keys are, or how parameter FQNs are constructed.  I think we should make sure the API is completely generic so we don't have to debug problems later on where something got changed higher up and ended up breaking undocumented assumptions in the checkpoint storage code.

kumpera commented on torch/distributed/_shard/checkpoint/filesystem.py:
```diff
@@ -0,0 +1,127 @@
+            return
+
+        with open(os.path.join(self.path, ".metadata"), "wb") as metadata_file:
+            pickle.dump(metadata, metadata_file)
```
We're using `torch.save` for saving individual entries, which uses pickle as well, that's another problematic place.

We're considering switching to flatbuffers, but the discussion still going on.

simpkins replied:
For tensors I think we need to write out the tensor contents directly.  We want to minimize the I/O overhead here, and therefore don't want to go through a serialization layer that might force copying the data.  For tensors stored in GPU we also should write out the data in a streaming fashion: we ideally should not have to hold the entire tensor data in CPU memory at once while saving the checkpoint.

For tensor metadata (i.e., the tensor options) then we will need some serialization layer to save this data.  I would perhaps be more inclined to do this serialization in C++ rather than in Python.

kumpera commented on torch/distributed/_shard/checkpoint/filesystem.py:
```diff
@@ -0,0 +1,127 @@
+        if dist.is_initialized() and dist.get_rank() != 0:
+            return
+
+        with open(os.path.join(self.path, ".metadata"), "wb") as metadata_file:
```
I'm disabling state_dicts with a `.metadata` key for now until the whole storage_key -> file name discussion is settled.

kumpera commented on torch/distributed/_shard/checkpoint/state_dict_saver.py:
```diff
@@ -0,0 +1,150 @@
+    return (metadata, storage_keys, bytes_write_requests, tensor_write_requests)
+
+
+def save_state_dict(
```
Good point, tracking in https://github.com/pytorch/pytorch/issues/76451

kumpera commented on torch/distributed/_shard/checkpoint/storage.py:
```diff
@@ -0,0 +1,89 @@
+    """
+
+    @abc.abstractmethod
+    def write_bytes(self, requests: List[BytesWriteRequest]) -> Future[None]:
```
The current semantics is that it's the user responsibility to ensure the destination doesn't have files with the same name as the saved ones. I'll update the docs to reflect that.

Later on we can let users control overwrite semantics

simpkins replied:
I think it's a bad API choice to not check the user input here, and silently create garbage checkpoints if this happens and still report success at the end.

pritamdamania87 replied:
I do agree with Adam here, I think throwing an error in such cases seems like a much safer option and we should clearly document this contract in the interface.

simpkins commented on torch/distributed/_shard/checkpoint/filesystem.py:
```diff
@@ -0,0 +1,142 @@
+        self.path = path
+        os.makedirs(self.path, exist_ok=True)
+
+    def _check_for_bad_keys(self, requests):
+        if any([r for r in requests if r.storage_key == ".metadata"]):
+            raise ValueError("Cannot use key named '.metadata'")
```
It feels wrong to allow some StorageWriter implementations to only implement a subset of the API.  It seems like it will be confusing if checkpoint saving can fail with some StorageWriter implementations but succeed with others.

In general I think it would be better to just consistently ensure that we can support any arbitrary keys.  If we really want to reject some keys then it seems like we should do that consistently at a higher layer in the code, so that we don't have different behavior across different StorageWriters.

Note that with the current FileSystemWriter implementation we should also be rejecting any keys containing `/`  (and depending on the platform, possibly `\`).  We also probably can't handle the empty string, `.`, or `..` correctly.

kumpera replied:
Good point. I implemented a simple FQN -> storage key mapping that should be enough to enable downstream users to start experimenting with this API.

This mapping schema only your concerns around FileSystemWriter and is not generally useful beyond it. I'll work on a design to address it after this PR.

simpkins commented on torch/distributed/_shard/checkpoint/state_dict_saver.py:
```diff
@@ -0,0 +1,150 @@
+    storage_writer.write_metadata(metadata=metadata)
+    bytes_futures = storage_writer.write_bytes(bytes_write_requests)
+    tensor_futures = storage_writer.write_tensors(tensor_write_requests)
+    bytes_futures.wait()
+    tensor_futures.wait()
```
So just to confirm, for now if `save_state_dict()` throws an exception the code may still continue writing out data for some indeterminate amount of time in the future?

Are there any thread safety concerns if a StorageWriter implementation may still be running in another thread and reading from tensors or other objects that the user passed in?

kumpera replied:
While we now properly wait for all futures, this is hardly enough to properly handle failure.

Failure remains local to a rank which, without explicit user coordination, will lead to the whole job hanging due to SPMD desync.

simpkins commented on torch/distributed/_shard/checkpoint/state_dict_loader.py:
```diff
@@ -0,0 +1,234 @@
+
+    bytes_futures.wait()
+
+    # Addtional steps are required to convert the bytes to its original type
+    # Note that this is NOT inplace,
+    # it creating a new object and replace what's in the state dict
+    for req in bytes_read_requests:
+        fqn = req.storage_key
+        # Ensure the BytesIO is rewound
+        req.bytes.seek(0)
+        state_dict[fqn] = torch.load(req.bytes)
```
What about defining 2 layers of APIs here: a lower-level API that requires `BytesIO`, and then a higher-level API on top of it that can accept anything and performs conversion to `BytesIO` on an as-needed basis by calling `torch.save()`?

For users that want to ensure backwards compatibility of checkpoints, and avoid possibly allowing arbitrary code execution vulnerabilities, this allows them to use the lower-level API and ensure that they cannot end up accidentally using `pickle`.

xunnanxu replied:
> The alternative design would require significantly more work than just switching from torch.save to save_state_dict which would be detrimental to our users a stifle adoption.

This arguably is not quite true. The general idea is users should be able to choose how they want serde to be done via `state_dict()` and `load_state_dict()` - they are defined in models. Today unfortunately a state_dict is `Dict[str, Any]` in pytorch but there's nothing preventing the checkpoint layer to say if you want to use it, please convert non tensor objects to `BytesIO` or otherwise I'll reject you.

For general APIs like this, IMHO it's always better to start restricted and then loosened over time. PyTorch generally tends to do the opposite and favor "user adoption" first which incurs hacks in lower level of the whole stack which arguably would lead to unresolvable tech debt over time.

kumpera replied:
The current API does what you suggest, which is to have the storage layer only deal with BytesIO and Tensor.

What we lack is pluggable serialization such that we can avoid pickle or have a more efficient variant of `torch.save` for tensors.

Designing this pluggable piece is not trivial and will take some time. I'm tracking this issue in https://github.com/pytorch/pytorch/issues/76441 as a follow up change after this one.

pritamdamania87 commented on test/distributed/_shard/checkpoint/test_file_system_checkpoint.py:
```diff
@@ -0,0 +1,506 @@
+        self.emb = torch.nn.EmbeddingBag(5, 10)
+
+
+# The ShardedModels are borrowed from test/distributed/_sharded_tensor/test_sharded_tensor.py
+class MyShardedModel3(torch.nn.Module):
```
nit: Maybe move these to a common location and reuse? Ex: torch/testing/_internal/distributed/_shard/sharded_tensor/_test_st_common.py

kumpera replied:
👍

pritamdamania87 commented on torch/distributed/_shard/checkpoint/metadata.py:
```diff
@@ -0,0 +1,79 @@
+    storage_key: str
+
+    # Tensor sizes
+    size: Tuple[int, ...]
```
Use torch.Size here instead as the type to be inline with rest of PyTorch?

pritamdamania87 commented on torch/distributed/_shard/checkpoint/metadata.py:
```diff
@@ -0,0 +1,79 @@
+
+
+@dataclass
+class TensorStorageMetadata:
```
Is this actually sufficient in terms of metadata that we need? Do we need something like [TensorProperties](https://github.com/pytorch/pytorch/blob/ba205a7eae357135f937e0501f7748d2dd953eba/torch/distributed/_shard/sharded_tensor/metadata.py#L15) too?

For example `ShardedTensorMetadata` has all of this information. It seems like it would be useful to just query a checkpoint and get information about the tensor's shape and everything else as well.

kumpera replied:
Reusing TensorProperties would be ideal. But this is blocked on: https://github.com/pytorch/pytorch/pull/76679

pritamdamania87 commented on torch/distributed/_shard/checkpoint/state_dict_saver.py:
```diff
@@ -0,0 +1,150 @@
+
+    Args:
+        state_dict: The instance to plan for.
+        include_non_replicated_tensors: Include non-sharded tensors even if rank != 0
```
nit: We are calling it non-sharded in one place and non-replicated in another place. Lets have consistent terminology.

kumpera replied:
I replaced non-sharded to be more specific to non ShardedTensor items.

pritamdamania87 commented on torch/distributed/_checkpoint/storage_writer.py:
```diff
@@ -0,0 +1,107 @@
+    def __init__(self, path: str) -> None:
+        super().__init__()
+        self.path = path
+        os.makedirs(self.path, exist_ok=True)
```
Looks like we still need to address this issue?

kumpera commented:
@pritamdamania87 
> In that case, we can just have prepare_storage take the metadata as a parameter instead?
Regarding `prepare_storage`.

Metadata is actually not enough because it doesn't have any information telling which rank is saving what, making it quite error prone.

Pass a list of FQNs would be annoying for Shards since you'd need to scan through the list of ShardStorageMetadata.

Passing a list of storage keys is the most reasonable approach for now, given it's the best way to tie back to write calls.

I expect this method to completely go away based on a storage coalescing prototype.

wilson100hong commented on torch/distributed/_shard/checkpoint/metadata.py:
```diff
@@ -0,0 +1,79 @@
+@dataclass
+class BytesReadRequest:
+    bytes: io.BytesIO
+    storage_key: str
```
Good point. IIRC for now there is no usage for non-tensor resharding. But might be helpful for feature like "checkpoint per-rank data loaders buffer". In practical, dataloader has a buffer to store prefetched batches but not trained yet. When saving data loader checkpoint, we need to drain their buffers, otherwise we will loss batches in buffer when resume it (data loss).

If we also want to save these "per rank dataloader buffer" in checkpoint as bytes, a reshardable BytesReadRequest is helpful when number of dataloaders (== number of trainers) change after resume

kumpera replied:
That's sounds like an important scenario to cover, but quite far from the current design we currently have.
I'm taking notes of this so we can tinker with it.

wilson100hong commented on torch/distributed/_shard/checkpoint/resharding.py:
```diff
@@ -0,0 +1,290 @@
+)
+
+def _create_storage_key(
+        storage_key_to_fqn: Dict[str, str],
```
indention

xunnanxu commented on torch/distributed/_shard/checkpoint/storage.py:
```diff
@@ -0,0 +1,167 @@
+        pass
+
+    @abc.abstractmethod
+    def write_tensors(self, requests: List[TensorWriteRequest]) -> Future[None]:
```
The storage layer can write the requests w/ additional customizations (e.g. chunking/batching/quantization etc.) so it would be important for this layer to clarify how storage metadata can be persisted using the APIs provided.

Essentially there are 2 options:

**Option 1:**
change `write_tensors()` to return `List[T]` where `T` is the type of storage metadata (can be defined by `StorageWriter` implementation) and `write_metadata()` takes additional `List[T]` as storage metadata tied to the tensor metadata.

Pros:
a. No cross-rank communication required for `StorageWriter`. The caller is responsible for that if needed (or if something needs to be called in rank 0 only).
b. API layer is explicit about when `StorageWriter` should return the storage metadata.

Cons:
a. Introduction of `T` where `T` is to some extent freeform.

**Option 2:**
Make storage metadata handling an implementation detail inside a writer.

Example pseudo SPMD impl:

```
def write_tensors(self, requests: List[TensorWriteRequest]) -> Future[None]:
  storage_metadata = self._do_write(requests)
  self._local_storage_metadata = storage_metadata

def write_metadata(self, metadata: Metadata) -> None:
  storage_metadata_all_ranks = pg.gather(self._local_storage_metadata)
  if pg.my_rank == 0:
    full_metadata = self._transform(metadata, storage_metadata_all_ranks)
    self._do_write(full_metadata)
```

Pros:
1. No `T` introduced

Cons:
1. Storage layer needs to be communication aware so the layering may be unclear (e.g. there may need to be SPMD/non-SPMD version even though the per rank IO logic is exactly the same).
2. Caller contract change (e.g. for SPMD `write_metadata()` needs to be called in every rank not just rank 0)
3. Implicit metadata handling.

Given these, option 1 seems preferred.

kumpera replied:
@xunnanxu  There's a third option worth considering.

Rank0 computes a global plan and scatters it before read/write starts.

Pros:
- No reconciliation at the end.
- No collectives in storage layer

Cons:
- Storage layer needs to be able to plan for non-local rank

I believe Option 1 is better still.

On point (1). I assume `write_tensor` would return `List[StorageMetadata]` which, in turn, would be defined as the following:

```
class StorageMetadata:
  fqn: str #FQN was provided in write request
  storage_data: Any # object that is serializable by `torch.distributed.gather_object`
```

This is follow up work that should happen after this PR lands. I'm tracking it here: https://github.com/pytorch/pytorch/issues/76395

xunnanxu replied:
The option 3 you mentioned may not work for certain storage layers (or maybe it's me not understanding the plan and you could help clarify). I assume in a perfect world, the state dict saver can be reused. And S3/Azure blob can be plugged in a `StorageWriter` implementations. In that case, a specific storage implementation (say Linux fs) may not have file names but only file descriptors - this implementation clearly has its own quirks that cannot be foreseen by higher level abstractions, and it doesn't make sense to force it to honor "storage keys" from this proposal because it just can't. In that sense the most sensible way would be to let the `StorageWriter` layer determine the actual ids (FDs or whatever) and return those to caller to merge them into the final metadata.

xunnanxu commented on torch/distributed/_shard/checkpoint/storage.py:
```diff
@@ -0,0 +1,167 @@
+    4) (if rank == 0) write_metadata().
+
+    Rank 0 will wait for (2) and (3) before writing metadata.
+    The writes from (2) and (3) are initiated before any waiting is done.
```
Is 3 called after future in 2 completes? Seems a bit unclear.
What if 2 fails? What's the failure handling story?

kumpera replied:
Updated the doc. The idea is that all writes are waited together and failure of either will invalidate the checkpoint.

The error handling story is going to be properly flushed out after this PR as it needs to be done across all ranks to ensure the checkpoint is correct.

xunnanxu commented on torch/distributed/_shard/checkpoint/state_dict_saver.py:
```diff
@@ -0,0 +1,136 @@
+        bytes_write_requests,
+        tensor_write_requests,
+    ) = _prepare(state_dict)
+    storage_writer.prepare_storage(storage_keys=storage_keys)
+    storage_writer.write_metadata(metadata=metadata)
+    bytes_futures = storage_writer.write_bytes(bytes_write_requests)
+    tensor_futures = storage_writer.write_tensors(tensor_write_requests)
+    bytes_futures.wait()
+    tensor_futures.wait()
```
This order is inverted from the doc:
The API doc says:

```
    1) prepare_storage with a dict of storage_key -> size. This can be used to
        pre-allocate storage.
    2) write_bytes
    3) write_tensors.
    4) (if rank == 0) write_metadata().
```

This doesn't do that nor does it do the rank 0 check?
Also this seems to write metadata even if writing tensor fails, which seems ok but also a bit odd.

kumpera replied:
I changed the order to reflect the docs and make sure that failures are detectable.

xunnanxu commented on torch/distributed/_shard/checkpoint/filesystem.py:
```diff
@@ -0,0 +1,131 @@
+) -> Path:
+    return (Path(storage_path) / storage_key)
+
+class FileSystemWriter(StorageWriter):
```
This example seems a bit flimsy to certain extent. Even though it's just a prototype, we probably need something better than a toy:

1. It does not cover failure handling logic. Just a quick example: what's the handling logic if `torch.save()` fails? How does caller handle that?
2. This does not show the capability of `write_metadata()` and arguably it should - for example the file system writer should probably have an example to show how to handle Windows case-insensitive file system when model wants to save `foo` and `Foo` (how to prevent one from overwriting the other using storage metadata).
3. The reader/writer does not say "I'm an SPMD implementation" (e.g. injecting `pg` in constructor) yet they just assume that which may make it hard to support generic resharding impl (like async training using RPC).

kumpera replied:
1. Failure handing still under design and I plan to work on it next. Tracking it here: https://github.com/pytorch/pytorch/issues/76395

2. The current design uses generated unique storage_keys (it hashes keys) so there's no need to handle case-insensitive filesystems and I believe this is the way to go as it solves other issues such as invalid characters and FQNs that are relative paths (IE `../foo`).

3. If Reader / Writer needs to use collectives, they must assume they are operating over the root PG. The current setup supports resharding ShardedTensors just fine. I'm not sure what you mean by generic resharding in this context.

As we introduce collectives to load/save state_dict in order to coordinate saving we'll have to allow a PG to be passed in.

xunnanxu replied:
> The current design uses generated unique storage_keys (it hashes keys) so there's no need to handle case-insensitive filesystems and I believe this is the way to go as it solves other issues such as invalid characters and FQNs that are relative paths (IE ../foo).

So this is not how we internally handle storage metadata today and not having a plan to support that would be a hard blocker for adoption (in a sense that no production workload can support this OSS solution until there's a story).

Today say a param `foo` is sharded into 2 ranks with shard FQN `foo.1` and `foo.2`, the storage may further chunk each shard to 2 files, or store the 2 shards into 1 file for optimization (if the shards are small and stacked into the same rank somehow).

Let's say it's the former so effectively we would store:

```
{
  "foo": {
    "dtype": "int32",
    "shards": [
      "foo.1": {
        "sizes": [5],
        "chunks": [
          {
            "filename": "foo.1_1",
            "offset": 0,
            "length": 10,
          },
          {
            "filename": "foo.1_2",
            "offset": 0,
            "length": 10,
          }
        ]
      },
      "foo.2": {
        "sizes": [5],
        "chunks": [
          {
            "filename": "foo.2_1",
            "offset": 0,
            "length": 20,
          },
        ]
      }
    ]
  }
}
```

Could you explain how this case can be supported properly when it comes to metadata transformation?

3. If Reader / Writer needs to use collectives, they must assume they are operating over the root PG. The current setup supports resharding ShardedTensors just fine. I'm not sure what you mean by generic resharding in this context.

In this implementation, it's not clear what the reader/writer responsibility is (which is what @simpkins points out above). Clearly in the example impl, the `write_tensor()` part is called in every rank, while the `write_metadata()` part is not clear at all.

In the api doc, it says

```
    3) write_tensors.
    4) Wait for (2) and (3) futures. If either fail, abort checkpoint.
    5) (if rank == 0) write_metadata().
```
which indicates **the caller** is responsible for checking if this needs to be called in specific ranks, yet in the example, it's the implementation.

This just shows that the proposal itself is not clear about who is responsible for that.

By generic, what I meant was the implementation should not be bound to process group as that seems to assume the synchronization is done by collective only. This is not universally true. PyTorch itself doesn't say all training needs to be full-sync using SPMD. If synchronization is done via RPC, this assumption would break. In theory, the same storage writer should handle both cases with zero code change, so it's better for caller to ensure the rank 0 check + additional synchronization if any across hosts/ranks.

BTW, can we not resolve comments until there's a real resolution?

kumpera replied:
Oh, and I'm moving the `rank 0` check away from FileSystemWriter

xunnanxu commented on torch/distributed/_shard/checkpoint/metadata.py:
```diff
@@ -0,0 +1,90 @@
+
+
+# Metadata for each param.
+@dataclass
```
A bit confused what this `ShardedTensorStorageMetadata` is supposed to handle.

Ostensibly, each storage implementation may want to have its own specific details included (e.g. a file system based may want to add file name mapping and an S3 based may want to know the bucket name).

It's not possible for the torch.dist version to handle everything. To me torch.distributed should only define the high level logical tensor metadata contract (dtype, strides, sizes etc. for a shard) and nothing beyond that. Storage layer should be free to choose whatever additional metadata to be "fused" in and it's not restricted to be 1-1 (a shard `foo` could be chunked into 5 files for example so the loading logic should know the mapping between that and the 5 files).

kumpera replied:
This only features the metadata that's not specific of any storage implementation.

xunnanxu replied:
Maybe help us understand why is `ShardMetadata` placed in `ShardStorageMetadata`?
I understand why we need maybe `storage_key` or `length` but `ShardedTensorMetadata` already has a list of `ShardMetadata`. What's the singleton in storage metadata for?

kumpera replied:
The core issue here is that there's no guarantee on ordering or equivalency between the `ShardMetatada` list and the `ShardStorageMetadata` list.

We could replace that `ShardeMetadata` field with an index into the ShardedTensorMetadata.
It would complicate things a bit but avoid the redundant data.

pritamdamania87 replied:
One option would be to not use ShardedTensorMetadata at all but some of its internal properties, so something like this to avoid duplication:

```
@dataclass
class ShardedTensorStorageMetadata(object):
    shards_metadata_storage: List[ShardStorageMetadata] = field(default_factory=list)
    size: torch.Size = field(default=torch.Size([]))
    tensor_properties: TensorProperties = field(
        default=TensorProperties(dtype=torch.get_default_dtype(),
                                 layout=torch.strided,
                                 requires_grad=False,
                                 memory_format=torch.contiguous_format,
                                 pin_memory=False))
```

We could have simple methods to convert from this to ShardedTensorMetadata and vice-versa.

kumpera replied:
@pritamdamania87  that makes sense, specially once we switch to local planning.

kumpera commented on torch/distributed/_shard/checkpoint/metadata.py:
```diff
@@ -0,0 +1,90 @@
+@dataclass
+class Metadata:
+    # Metadata for the state dict.
+    # This includes the MD for Tensors and ShardedTensors. ByteIO objects are skipped
```
This comments becoming stale vis-a-via the implementation. Fixed.

kumpera commented:
@simpkins  I'm tracking future work to improve the serialization story here: https://github.com/pytorch/pytorch/issues/76441

kumpera commented on torch/distributed/_shard/checkpoint/storage.py:
```diff
@@ -0,0 +1,167 @@
+        pass
+
+    @abc.abstractmethod
+    def write_metadata(self, metadata: Metadata) -> None:
```
I'm updating the docs to be clear that `write_metadata` does commit the checkpoint.

On the second part of your comment, is this about a StorageWriter impl having freedom to use any schema/format for the data or whether any data needs to be written at all.

An example of my later point is a storage layer where each rank saves the metadata along the data they are writing and the loader consumes that to produce the global metadata object.

Is this sort of scenario that you had in mind?

While I agree with you that this needs to be addressed, this can be done in a follow up change. I tracking this in https://github.com/pytorch/pytorch/issues/76441

kumpera commented on torch/distributed/_shard/checkpoint/storage.py:
```diff
@@ -0,0 +1,89 @@
+        pass
+
+    @abc.abstractmethod
+    def read_tensors(self, requests: List[TensorReadRequest]) -> Future[None]:
```
Does this happen in practice? The recommended practice for param sharing is to reuse the param from a different module and not introduce storage aliasing.

The current behavior is undefined values because writes to the aliased storage are allowed to race between each other.

I agree that failing to load would be the most desirable outcome. We should address it on a followup PR as this doesn't happen with the models we're testing this with.

I filed this issue to track this: https://github.com/pytorch/pytorch/issues/76745

kumpera commented on torch/distributed/_shard/checkpoint/state_dict_saver.py:
```diff
@@ -0,0 +1,150 @@
+    `torch.distributed._shard.sharded_tensor.state_dict_hook` prior to
+    calling `state_dict()` on the top module.
+
+    There is no guarantees of Backwards Compatibility across PyTorch versions
+    for saved state_dicts.
```
Until we move away from torch.load / torch.save such guarantees are moot due to: https://pytorch.org/docs/stable/notes/serialization.html#saving-and-loading-scriptmodules-across-pytorch-versions

I agree that having a proper, stable envelope format is necessary and we must discuss it before we move this feature to a Beta feature.

I'm tracking this here: https://github.com/pytorch/pytorch/issues/76747

kumpera commented on torch/distributed/_shard/checkpoint/state_dict_loader.py:
```diff
@@ -0,0 +1,234 @@
+    return (bytes_read_requests, tensor_read_requests)
+
+
+def load_state_dict(
+    state_dict: Dict[str, Any],
+    storage_reader: StorageReader,
+) -> None:
```
Keys are available by doing `storage_reader.read_metadata().state_dict_metadata.keys()`

We have a `validate` method that takes a `state_dict` and a `Metadata` object as loaded from `read_metadata` that verifies checkpoint compatibility without doing any IO.

Once the design starts to settle down, the on disk format, in particular. I plan to work on a CLI tool to operate on checkpoints.

kumpera commented on torch/distributed/_shard/checkpoint/state_dict_loader.py:
```diff
@@ -0,0 +1,234 @@
+                    tensor=tensor,
+                    storage_key=fqn,
+                    offsets=tuple([0] * len(tensor.size())),
+                    lengths=md.size,
```
I have this removed on a follow up branch:  https://github.com/kumpera/pytorch/commit/f83dab205ede624e5ab13eba04ce59f6a29cd361#diff-f8ffc3f7ad564c1ec0a035f92696f2df0e5a65ba4296f042c2b66f10209326b3L56

kumpera commented on torch/distributed/_shard/checkpoint/filesystem.py:
```diff
@@ -0,0 +1,127 @@
+from .storage import StorageReader, StorageWriter
+
+class FileSystemWriter(StorageWriter):
+    def __init__(self, path: str) -> None:
+        super().__init__()
+        self.path = path
+        os.makedirs(self.path, exist_ok=True)
```
That's a good point.

I'm applying your suggested change plus adding a prepare/finish method to StorageWriter.

kumpera commented on torch/distributed/_shard/checkpoint/filesystem.py:
```diff
@@ -0,0 +1,127 @@
+                # for unblocking. There are two drawbacks:
+                # 1. `torch.save` is pickle based, and pickle is not known for its
+                #   compatibility, we should consider replacing it with a more
+                #   stable option.
```
Yes, I completely agree but I think this is an acceptable compromise to get this change in and then we can start working on coming up with an alternative. What do you think?

kumpera commented on torch/distributed/_shard/checkpoint/filesystem.py:
```diff
@@ -0,0 +1,127 @@
+                    (view_cached is not None and view_cached.device != req.tensor.device):
+
+                with open(os.path.join(self.path, req.storage_key), "rb") as storage:
+                    view_cached = cast(Tensor, torch.load(storage, map_location=req.tensor.device))
```
We only produce Tensor reads when the checkpoint has a tensor related entry.
The ways this could fail are either due to bugs in serialization or an adversarial checkpoint.

kumpera commented on torch/distributed/_shard/checkpoint/storage.py:
```diff
@@ -0,0 +1,89 @@
+        This blocking call can be overwritten by the subclass.
+        It can use `storage_keys` to plan for any write preformace optimization.
+        e.g. non sequential and parallel writes.
+        By default, it does nothing
```
This function was originally intended for your usecase actually.

It's a vestigial part of the design that will go away once we figure out the right interface for StorageWriters with complex writing logic.

kumpera commented on torch/distributed/_shard/checkpoint/resharding.py:
```diff
@@ -0,0 +1,198 @@
+    shard: ShardMetadata
+) -> str:
+    shard_key = "_".join([str(i) for i in shard.shard_offsets])
+    return f"{tensor_storage_key}{STORAGE_KEY_SEPARATOR}{shard_key}"
```
While this could generate duplicate FQNs, we ensure there's no conflict by collecting all FQNs, hashing them and use a deterministic way to resolve collisions.

You can see that in _create_storage_key.

kumpera commented on torch/distributed/_shard/checkpoint/resharding.py:
```diff
@@ -0,0 +1,198 @@
+    tensor_storage_key: str,
+    shard: ShardMetadata
+) -> str:
+    shard_key = "_".join([str(i) for i in shard.shard_offsets])
```
👍

kumpera commented on torch/distributed/_shard/checkpoint/state_dict_saver.py:
```diff
@@ -0,0 +1,150 @@
+    return (metadata, storage_keys, bytes_write_requests, tensor_write_requests)
+
+
+def save_state_dict(
```
I agree we want to enable save/load time transformation.

I believe before we introduce a fully fledged API, we should support the specific case of quantization to understand how transforms needs to be done.

kumpera commented on torch/distributed/_shard/checkpoint/filesystem.py:
```diff
@@ -0,0 +1,127 @@
+    def write_metadata(self, metadata: Metadata) -> None:
+        # Once write metadata once as each Metadata has the global view
+        if dist.is_initialized() and dist.get_rank() != 0:
+            return
```
Agreed, the storage layer must not be aware of torch.distributed. It should work without it.

Fixed the code.

kumpera commented on torch/distributed/_shard/checkpoint/state_dict_saver.py:
```diff
@@ -0,0 +1,150 @@
+                metadata.state_dict_metadata[fqn] = _compute_tensor_md(fqn, obj)
+        else:
+            bytes_io = io.BytesIO()
+            torch.save(obj, bytes_io)
```
We have two alternatives here:

Option (1)
Require users to serialize upfront to BytesIO. Fair if run into anything else.

Pros:
- don't have to deal with object serialization
- simpler implementation

Cons:
- API is less usable and its behavior is further diverging from torch.load / torch.save.
- It's harder to implement it efficiently and avoid the intermediate copy to a ByteIO

Option (2)
Serialize all non tensor objects.

Pros:
- Simpler API, easier to use
- Possibly more efficient (but likely with stuff that is not perf relevant)

Cons:
- This requires us to use pickle.

kumpera commented on torch/distributed/_shard/checkpoint/state_dict_saver.py:
```diff
@@ -0,0 +1,150 @@
+        elif isinstance(obj, Tensor):
+            # The assumption is that non ShardedTensors are full replicated across all ranks
+            # So we just need one from Rank 0.
+            # If that's not the case, we will update later.
```
The wording here is pretty contrived. We handle 3 kinds of objects: ShardedTensors, Tensors and BytesIO. By BytesIO I mean everything else that we currently `torch.save` to a BytesIO. We save data this way:

On all ranks:
save local shard of all ShardedTensors

On rank0 only:
save Tensors and BytesIO objects

There are no provisions to save anything but sharded tensors on rank > 0.

kumpera replied:
The wording here is pretty contrived. We handle 3 kinds of objects: ShardedTensors, Tensors and BytesIO. By BytesIO I mean everything else that we currently `torch.save` to a BytesIO. We save data this way:

On all ranks:
save local shard of all ShardedTensors

On rank0 only:
save Tensors and BytesIO objects

There are no provisions to save anything but sharded tensors on rank > 0.

kumpera commented on torch/distributed/_shard/checkpoint/state_dict_saver.py:
```diff
@@ -0,0 +1,150 @@
+            )
+            bytes_write_requests.append(bwr)
+
+    storage_keys: Dict[str, int] = {
+        req.storage_key: req.tensor.nelement() * req.tensor.element_size()
+        for req in tensor_write_requests
+    }
```
good point.

kumpera commented on torch/distributed/_shard/checkpoint/state_dict_loader.py:
```diff
@@ -0,0 +1,234 @@
+                raise ValueError(
+                    f"Invalid checkpoint metadata for {fqn}, " +
+                    "expected TensorStorageMetadata but found {type(md)}"
+                )
```
Here's a follow up branch that implements it: https://github.com/kumpera/pytorch/commit/f83dab205ede624e5ab13eba04ce59f6a29cd361

pritamdamania87 commented:
> Metadata is actually not enough because it doesn't have any information telling which rank is saving what, making it quite error prone.

> Pass a list of FQNs would be annoying for Shards since you'd need to scan through the list of ShardStorageMetadata.

> Passing a list of storage keys is the most reasonable approach for now, given it's the best way to tie back to write calls.

> I expect this method to completely go away based on a storage coalescing prototype.

Ah my bad, I actually meant to say passing a `List[WriteRequests]` to `prepare_storage` so that it is more generic and has all the information about future incoming writes. But if this is probably going to change quite a bit based on the coalescing prototype we can revisit this in a follow up PR.

pritamdamania87 commented on torch/distributed/_shard/checkpoint/resharding.py:
```diff
@@ -0,0 +1,290 @@
+
+def _prepare_sharded_tensor_write(
+    sharded_tensor: ShardedTensor,
+    storage_key: str,
```
nit: This should be `fqn`? It was a bit confusing while reading the code, where I thought this is the storage_key in the StorageMetadata but it is actually the FQN of the parameter?

pritamdamania87 commented on torch/distributed/_shard/checkpoint/resharding.py:
```diff
@@ -0,0 +1,290 @@
+
+    for shard in sharded_tensor.local_shards():
+        tensor = shard.tensor.detach()
+        shard_storage_key = shard_to_storage_key[_get_shard_key(shard.metadata)]
```
Why do we need the loop above? Can't we just call `_get_shard_storage_key` here and eliminate the loop above?

pritamdamania87 commented on torch/distributed/_shard/checkpoint/state_dict_saver.py:
```diff
@@ -0,0 +1,159 @@
+        if isinstance(obj, ShardedTensor):
+            st_write_reqs, st_md = _prepare_sharded_tensor_write(obj, fqn, storage_key_to_fqn)
+            tensor_write_requests += st_write_reqs
+            metadata.state_dict_metadata[fqn] = st_md
```
I'm assuming we plan to change this where only rank 0 computes the global metadata and not every rank?

kumpera replied:
Yes, the plan is to gather each rank metadata after writes.

pritamdamania87 commented on torch/distributed/_shard/checkpoint/resharding.py:
```diff
@@ -0,0 +1,290 @@
+            storage_key=shard_storage_key,
+        )
+        write_requests.append(wr)
+    return write_requests, _compute_sharded_tensor_md(
```
`_compute_sharded_tensor_md` seems to be computing the metadata for the entire Tensor although here we are writing only local shards. It seems like the metadata should be computed only once on rank 0 and things like `_prepare_sharded_tensor_write` should only care about the local writes?

kumpera replied:
The current design mixes the creation of Metadata and write requests such that splitting them would be quite tricky.

I don't think it's worth refactoring this given we'll replace it in the near future.

pritamdamania87 commented on torch/distributed/_shard/checkpoint/state_dict_saver.py:
```diff
@@ -0,0 +1,159 @@
+            # This produces incomplete MD for rank > 0 since we won't populate bytes_io.
+            # This is ok since only rank == 0 uses this data
+            if write_replicated_data:
+                torch.save(obj, bytes_io)
```
I guess there should be a way for the user to override this if they don't want to use pickle.

kumpera replied:
Yes, the plan is to introduce a layer that will sidesteps all issues around pickling.

pritamdamania87 commented on torch/distributed/_shard/checkpoint/state_dict_loader.py:
```diff
@@ -0,0 +1,234 @@
+                    "expected ShardedTensorStorageMetadata but found {type(md)}"
+                )
+        elif isinstance(obj, torch.Tensor):
+            tensor = obj.detach()
```
Maybe add a comment here explaining why we need the detach() and the fact that it is not in-place so it doesn't actually affect the Tensor in state_dict.

pritamdamania87 commented on test/distributed/_shard/checkpoint/test_file_system_checkpoint.py:
```diff
@@ -0,0 +1,384 @@
+        dist.barrier()
+
+
+class TestDistributedReshardOnLoad(ShardedTensorTestBase):
```
My understanding is that this tests resharding with same world_size? Do we have a unit test that verifies that we can change the world_size as well and load?

kumpera replied:
Yes, we don't have any tests using different world sizes though we can get a similar effect by changing the sharding spec.

I thought about on how to do it but it would require to spawn multiple sets of process groups, one for each world size and that sounded like over-engineering. Am I missing some simple way to do it?

kumpera committed:
workaround ST:gather hang on CI by using a bootleg version.

pritamdamania87 reviewed (approved):
Stamping this PR to unblock some additional iterative work to polish out some of the implementation and APIs going forward.

The understanding here is that the concerns raised by @xunnanxu, @simpkins and other reviewers will all be addressed in an iterative manner after this PR lands since adding everything to this PR would be too hard to manage and review.

kumpera commented on torch/distributed/_shard/checkpoint/state_dict_saver.py:
```diff
@@ -0,0 +1,159 @@
+        tensor_write_requests,
+    ) = _prepare(state_dict)
+
+    is_rank0 = not dist.is_initialized() or dist.get_rank(process_group) == 0
```
This API is meant to function without PG initialized when used with a single GPU.

kumpera commented on torch/distributed/_shard/checkpoint/storage.py:
```diff
@@ -0,0 +1,198 @@
+        pass
+
+
+    def prepare_storage(self, storage_keys: Dict[str, int]) -> None:
```
good point.

kumpera commented on torch/distributed/_shard/checkpoint/metadata.py:
```diff
@@ -0,0 +1,90 @@
+    # Keys are the same from the `state_dict` used.
+    state_dict_metadata: Dict[str, STORAGE_TYPES]
+
+    def __getstate__(self) -> bytes:
+        serialized = pickle.dumps(self.state_dict_metadata)
+        return serialized
+
+    def __setstate__(self, state: bytes) -> None:
+        self.state_dict_metadata = pickle.loads(state)
```
It's not even needed at all, I'll just remove it.

kumpera commented on torch/distributed/_shard/checkpoint/storage.py:
```diff
@@ -0,0 +1,198 @@
+        """
+        Initialize storage to receive the checkpoint.
+
+        This method will only be called on rank0.
```
Yeah, that makes it clearer.

kumpera committed:
Cleanup documentations and remove redundant pickle code.

kumpera commented:
@pytorchbot merge this

janeyx99 commented:
> Merge failed due to Can't fetch all PR reviews Raised by https://github.com/pytorch/pytorch/actions/runs/2275691136

@osalpekar @malfet This is failing because there are 109 review comments on this PR but we only fetch the first 100. This could be solved with a similar concept as how we fetch more comments/check_runs.

janeyx99 commented:
On a side note, has the test_fsdp_clip_grad_norm_norm_type_2_0_nested_fsdp_False_cpu_offload_CPUOffload failure on the distributed test first shard of this PR been addressed?
